### PR TITLE
Add a factory for creating MesosSchedulerDrivers.

### PR DESCRIPTION
**Pull request by nickbp**
Each framework should be constructing these in the same way. Can be expanded as things are normalized further.

### CODE REVIEW

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  public SchedulerDriver create(
+      final Scheduler scheduler,
+      final FrameworkInfo frameworkInfo,
+      final String masterZkUrl,
+      final byte[] credentialSecret) {
+    LOGGER.info("Creating MesosSchedulerDriver for "
+        + "scheduler = {}, frameworkInfo = {}, masterZkUrl = {}, credentialSecret = {}",
+        scheduler, frameworkInfo, masterZkUrl, credentialSecret);
+
+    if (!frameworkInfo.hasPrincipal() || StringUtils.isEmpty(frameworkInfo.getPrincipal())) {
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Requiring a principal is ok, as Mesos allows you to make one up.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      throw new IllegalArgumentException(
+          "Unable to create MesosSchedulerDriver, FrameworkInfo lacks a principal: "
+              + frameworkInfo.toString());
+    }
+
+    Credential.Builder credentialBuilder = Credential.newBuilder()
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: However, always generating credentials shouldn't be a requirement.  In an unauthenticated environment, I'm not sure if providing unneeded credentials will result in everything being ok, or in things not working.  Better to not generate credentials when we expect them not to be used.

**Comment by nickbp** [comment]
Had another chat with security folks. Turns out the behavior should indeed be different than what's laid out here. Dropping assignee while I work on it.

**Comment by nickbp** [comment]
...aand we're back. PTAL.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  private static final Logger LOGGER =
+    LoggerFactory.getLogger(SchedulerDriverFactory.class);
+
+  /**
+   * If this environment variable is present in the scheduler environment, the master is using
+   * some form of sidechannel auth. When this environment variable is present, we should always
+   * provide a {@link Credential} with (only) the principal set.
+   */
+  private static String SIDECHANNEL_AUTH_MAGIC_ENVNAME = "DCOS_SERVICE_ACCOUNT_CREDENTIAL";
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Magic has bad connotations.  Please could you rename this variable?

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  /**
+   * Creates and returns a new {@link SchedulerDriver} without a credential secret.
+   *
+   * @param scheduler The Framework {@link Scheduler} implementation which should receive callbacks
+   *     from the {@link SchedulerDriver}
+   * @param frameworkInfo The {@link FrameworkInfo} which describes the framework implementation.
+   *     The 'principal' field MUST be populated and non-empty
+   * @param masterZkUrl The URL of the currently active Mesos Master, of the form "zk://host/mesos"
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Could we please rename masterZkUrl to something less implementation specific.  Zookeeper is potentially a transient implementation detail.  Maybe `masterUrl`.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    } else {
+      return new MesosSchedulerDriver(scheduler, frameworkInfo, masterZkUrl, credential);
+    }
+  }
+
+  /**
+   * Returns whether it appears that sidechannel auth should be used when creating the
+   * SchedulerDriver. Broken out into a separate function to customize behavior in tests.
+   */
+  protected boolean isSideChannelActive() {
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: I hate to carp on about names, but "SideChannel" implies illegitimacy.  I think this is a totally legitimate authentication mechanism which is being supported...  Don't have a good name suggestion however.  Naming is hard.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      LOGGER.info("Creating secret authenticated MesosSchedulerDriver for "
+          + "scheduler = {}, frameworkInfo = {}, masterZkUrl = {}, credentialSecret = {}",
+          scheduler, frameworkInfo, masterZkUrl, credentialSecret);
+    } else if (isSideChannelActive()) {
+      // Sidechannel auth is enabled. Provide a Credential with only the Principal set.
+      if (!frameworkInfo.hasPrincipal() || StringUtils.isEmpty(frameworkInfo.getPrincipal())) {
+        throw new IllegalArgumentException(
+            "Unable to create MesosSchedulerDriver for sidechannel auth, "
+                + "FrameworkInfo lacks required principal: " + frameworkInfo.toString());
+      }
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Lines 82-86 are a copy of 68-72.  Please, factor out into a function.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            "Unable to create MesosSchedulerDriver for secret auth, "
+                + "FrameworkInfo lacks required principal: " + frameworkInfo.toString());
+      }
+      credential = Credential.newBuilder()
+          .setPrincipal(frameworkInfo.getPrincipal())
+          .setSecretBytes(ByteString.copyFrom(credentialSecret))
+          .build();
+      LOGGER.info("Creating secret authenticated MesosSchedulerDriver for "
+          + "scheduler = {}, frameworkInfo = {}, masterZkUrl = {}, credentialSecret = {}",
+          scheduler, frameworkInfo, masterZkUrl, credentialSecret);
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Please put the log message before the attempt to create the Credential object.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        throw new IllegalArgumentException(
+            "Unable to create MesosSchedulerDriver for sidechannel auth, "
+                + "FrameworkInfo lacks required principal: " + frameworkInfo.toString());
+      }
+      credential = Credential.newBuilder()
+          .setPrincipal(frameworkInfo.getPrincipal())
+          .build();
+      LOGGER.info("Creating sidechannel authenticated MesosSchedulerDriver for "
+          + "scheduler = {}, frameworkInfo = {}, masterZkUrl = {}",
+          scheduler, frameworkInfo, masterZkUrl);
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Please put the log message before the attempt to create the Credential object.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    } else {
+      // No auth. Provide no credential.
+      credential = null;
+      LOGGER.info("Creating unauthenticated MesosSchedulerDriver for "
+          + "scheduler = {}, frameworkInfo = {}, masterZkUrl = {}",
+          scheduler, frameworkInfo, masterZkUrl);
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Please put the log message before the attempt to create the Credential object.

**Comment by gabrielhartmann** [comment]
Test coverage looks good, and mostly minor comments regarding some possible improvements.  @adam-mesos:  Logic look good to you?

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 9 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Done

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 8 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Done

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 7 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Done

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 6 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Done

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 5 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: This name is intended to encompass both Bouncer-based Framework Auth as well as potentially other auth in the future which would work via the same mechanism (eg Kerberos). Added a note to the comment explaining this.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 4 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Done

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 3 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Dropped "MAGIC" from the name

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      final FrameworkInfo frameworkInfo,
+      final String masterUrl,
+      final byte[] credentialSecret) {
+    Credential credential;
+    if (credentialSecret != null && credentialSecret.length > 0) {
+      // User has manually provided a Secret. Provide a Credential with Principal + Secret.
+      // (note: we intentionally avoid logging the content of the credential secret, just in case)
+      LOGGER.info("Creating secret authenticated MesosSchedulerDriver for "
+          + "scheduler[{}], frameworkInfo[{}], masterUrl[{}], credentialSecret[{} bytes]",
+          scheduler, frameworkInfo, masterUrl, credentialSecret.length);
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: I'd like to see the principal in this log message too, but I guess it's included in the frameworkInfo?

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * Extracts the Principal name from the provided {@link FrameworkInfo}, or throws an
+   * {@link IllegalArgumentException} (mentioning the provided {@code authType}) if the Principal
+   * is unavailable.
+   */
+  private static String getPrincipal(final FrameworkInfo frameworkInfo, final String authType) {
+    if (!frameworkInfo.hasPrincipal() || StringUtils.isEmpty(frameworkInfo.getPrincipal())) {
+      throw new IllegalArgumentException(
+          "Unable to create MesosSchedulerDriver for " + authType + " auth, "
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: `authType` seems irrelevant to this getter. Seems like better practice would be for the caller to catch the exception/error/empty-string and generate its own error message.
Similarly, this standalone static method shouldn't need to know about the MesosSchedulerDriver, or that its caller is trying to create one.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    return create(scheduler, frameworkInfo, masterUrl, null /* credentialSecret */);
+  }
+
+  /**
+   * Creates and returns a new {@link SchedulerDriver} with the provided credential secret.
+   *
+   * @param scheduler The Framework {@link Scheduler} implementation which should receive callbacks
+   *     from the {@link SchedulerDriver}
+   * @param frameworkInfo The {@link FrameworkInfo} which describes the framework implementation.
+   *     The 'principal' field MUST be populated and non-empty
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: You say MUST, and yet you don't assert or check/error.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  /**
+   * Returns whether it appears that side channel auth should be used when creating the
+   * SchedulerDriver. Side channel auth may take the form of Bouncer-based framework authentication
+   * or potentially other methods in the future (eg Kerberos). Broken out into a separate function
+   * to allow easy customization in tests.
+   */
+  protected boolean isSideChannelActive() {
+    return System.getenv(SIDECHANNEL_AUTH_ENV_NAME) != null;
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: I'll reiterate that my gut prefers an explicit "--authenticate" flag to decide whether to pass a Credential to the MesosSchedulerDriver, and then the "--secret_file" flag can determine whether the Credential will have a secret.
It seems weird to hardcode a magic "sidechannel" environment variable that would be different if trying to use the classic Kerberos authenticatee module, and could change names in the future (no promises).
But dcos-commons isn't my library, so I'll leave the final decision up to you guys.

**Comment by adam-mesos** [comment]
Logic looks good, but I'm still wary of basing the decision to include a Credential on a magic environment variable name instead of an explicit flag.

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 10 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: Yes this just dumps the whole info which contains the principal

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 11 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: This is a private utility method called in two places elsewhere in the class. The authType field is used internally for logging

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 12 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: The check is in getPrincipal()

File: src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java (hunk 13 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: I really don't want to punish users on auth-enabled clusters with manually having to check a box each time they install anything on their cluster. Meanwhile the envvar name is already going to be independently synchronized between other independent DCOS components, it's already semi-stable. This library only targets DCOS.

Comment (nickbp) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: And to add: due to how Universe works, each framework would also need to independently define that flag in their packaging, leaving a burden on the framework developer and leaving open a window for inconsistencies in how the flag is defined

Given both of the above, having the library deal with it automatically feels like the preferable option

Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java]: I agree with Nick.

**Pull request merged**